src/main/java/com/iota/iri/service/snapshot/impl/SnapshotServiceImpl.java

```diff
@@ -113,30 +118,43 @@ public void replayMilestones(Snapshot snapshot, int targetMilestoneIndex) throws
                 if (currentMilestone != null) {
                     StateDiffViewModel stateDiffViewModel = StateDiffViewModel.load(tangle, currentMilestone.getHash());
                     if(!stateDiffViewModel.isEmpty()) {
-                        snapshot.applyStateDiff(new SnapshotStateDiffImpl(stateDiffViewModel.getDiff()));
+                        stateDiffViewModel.getDiff().forEach((address, balance) -> {
+                            if (balanceChanges.computeIfPresent(address, (hash, value) -> balance + value) == null) {
+                                balanceChanges.putIfAbsent(address, balance);
+                            }
+                        });
                     }
 
-                    snapshot.setIndex(currentMilestone.index());
-                    snapshot.setHash(currentMilestone.getHash());
+                    lastAppliedMilestone = currentMilestone;
+                } else {
+                    skippedMilestones.add(currentMilestoneIndex);
+                }
+            }
 
-                    TransactionViewModel currentMilestoneTransaction = TransactionViewModel.fromHash(tangle,
-                            currentMilestone.getHash());
+            if (lastAppliedMilestone != null) {
+                try {
+                    snapshot.lockWrite();
 
-                    if(currentMilestoneTransaction != null &&
-                            currentMilestoneTransaction.getType() != TransactionViewModel.PREFILLED_SLOT) {
+                    snapshot.applyStateDiff(new SnapshotStateDiffImpl(balanceChanges));
 
-                        snapshot.setTimestamp(currentMilestoneTransaction.getTimestamp());
+                    snapshot.setIndex(lastAppliedMilestone.index());
+                    snapshot.setHash(lastAppliedMilestone.getHash());
+
+                    TransactionViewModel milestoneTransaction = TransactionViewModel.fromHash(tangle,
+                            lastAppliedMilestone.getHash());
+                    if(milestoneTransaction.getType() != TransactionViewModel.PREFILLED_SLOT) {
+                        snapshot.setTimestamp(milestoneTransaction.getTimestamp());
                     }
-                } else {
-                    snapshot.addSkippedMilestone(currentMilestoneIndex);
+
+                    for (int skippedMilestoneIndex : skippedMilestones) {
+                        snapshot.addSkippedMilestone(skippedMilestoneIndex);
```

Comment:
maybe add an `addAll()` method which may have more efficient functionality?

Reply:
This is just to provide support for old databases - the patched COO doesnt skip milestones anymore - so don't know if this is worth adding - actually this whole skippedMilestone stuff should be deprecated in a 2-3 months but while migrating from an existing DB we still need it.